lib/vector-common/src/internal_event/mod.rs

```diff
@@ -224,3 +245,38 @@ macro_rules! registered_event {
         }
     };
 }
+
+#[derive(Clone)]
+pub struct Cached<Tags, Event, Register> {
+    cache: Arc<RwLock<BTreeMap<Tags, Event>>>,
+    register: Register,
+}
+
+impl<Tags, Event, Register, Data> Cached<Tags, Event, Register>
+where
+    Data: Sized,
+    Register: Fn(&Tags) -> Event,
```

Comment:
This register function should likely be co-located with the event and not the cache. Could it be made a trait on the event type, potentially generated by a macro (maybe the same `registered_event` thing)?